File: Problems/String/1786.py
import sys

ans = []


# sys.stdin = open("input.txt", "r")


def kmp(txt: str, pat: str) -> int:
    pt = 1
    pp = 0
    skip = [0] * (len(pat) + 1)

    skip[pt] = 0
    while pt != len(pat):
        if pat[pt] == pat[pp]:
            pt += 1
            pp += 1
            skip[pt] = pp
        elif pp == 0:
            pt += 1
            skip[pt] = pp
        else:
            pp = skip[pp]

    pt = pp = 0
    while pt != len(txt) and pp != len(pat):
        if txt[pt] == pat[pp]:
            pt += 1
            pp += 1
        elif pp == 0:
            pt += 1
        else:
            pp = skip[pp]

        if pp == len(pat):
            ans.append(pt - pp + 1)
            pp = skip[pp]

    return pt - pp if pp == len(pat) else -1


s1 = input()
s2 = input()
idx = kmp(s1, s2)
print(len(ans))
for i in ans:
    print(i)

File: Problems/KMP/1786.cpp
#define _CRT_SECURE_NO_WARNINGS
#include <iostream>
#include <vector>
#include <string>

using namespace std;

vector<int> skipTable;
vector<int> answer;
string t;
string p;

void SetSkipTable()
{
	int prefix = 0;
	for (int i = 1, len = p.length(); i < len; i++)
	{
		while (prefix > 0 && p[i] != p[prefix])
		{
			prefix = skipTable[prefix - 1];
		}
		if (p[prefix] == p[i])
		{
			skipTable[i] = ++prefix;
		}
	}
}

void Kmp()
{
	int pat = 0, txt = 0;
	while (pat < p.length() && txt < t.length())
	{
		if (p[pat] == t[txt])
		{
			pat++;
			txt++;
		}
		else if (pat == 0)
		{
			txt++;
		}
		else
		{
			pat = skipTable[pat - 1];
		}

		if (pat == p.length())
		{
			answer.push_back(txt - pat + 1);
			pat = skipTable[pat - 1];
		}
	}
}

int main()
{
	ios::sync_with_stdio(false);
	cin.tie(nullptr);
	freopen("input.txt", "r", stdin);

	getline(cin, t);
	getline(cin, p);

	skipTable.assign(p.length(), 0);
	SetSkipTable();

	Kmp();

	cout << answer.size() << "\n";
	for (auto e : answer)
	{
		cout << e << " ";
	}

	return 0;
}